refactored ActivityService logging strategy

jyoo980 opened the pull request:
### Before submitting a pull-request to GitBucket I have first:

- [x] read the [contribution guidelines](https://github.com/gitbucket/gitbucket/blob/master/.github/CONTRIBUTING.md)
- [x] rebased my branch over master
- [x] verified that project is compiling
- [x] verified that tests are passing
- [x] squashed my commits as appropriate *(keep several commits if it is relevant to understand the PR)*
- [ ] [marked as closed using commit message](https://help.github.com/articles/closing-issues-via-commit-messages/) all issue ID that this PR should correct

This is a follow up from the conversation @takezoe and I had in [this conversation](https://github.com/gitbucket/gitbucket/pull/2523#discussion_r476191217)

jyoo980 commented on src/main/scala/gitbucket/core/service/ActivityService.scala:
```diff
@@ -97,478 +98,6 @@ trait ActivityService {
-
-  private def cut(value: String, length: Int): String =
-    if (value.length > length) value.substring(0, length) + "..." else value
+  def recordActivity[T <: BaseActivityInfo](info: T): Unit =
```
This is honestly the main benefit of this work. We no longer need to create a new function each time we want to log something. This function takes care of them all!

jyoo980 commented on src/main/scala/gitbucket/core/service/HandleCommentService.scala:
```diff
@@ -29,25 +36,31 @@ trait HandleCommentService {
+              updateClosed(owner, name, issue.issueId, false)
+          }
+
+          val (action, _) = actionOpt
```
I could definitely simplify this further, but a combination of not being too familiar with this specific part, and being tired from this refactor is making me put this off until later.

takezoe replied:
Okay, I got it. 😅

takezoe commented on src/main/scala/gitbucket/core/model/activity/comment/CommentActivityInfo.scala:
```diff
@@ -0,0 +1,95 @@
+import gitbucket.core.model.Profile.currentDate
+import gitbucket.core.model.activity.BaseActivityInfo
+
+sealed trait CommentActivityInfo extends BaseActivityInfo {
```
Even if we have a base trait for specific types of activity information, each sub case classes need to have conductor arguments and we need to implement conversion to `Activity` in each class. Therefore, these base trait would be not much meaningful.

jyoo980 replied:
Right, I understand what you're saying here. 

I think the benefit of this would be future ergonomics and extensibility. An alternative way of doing this would be to
* have each case class extend the `BaseActivityInfo` trait

This could remove the need for our base traits. That being said, I don't think this is the correct approach. I could see a point in the future where we want to have a function that consumes some activity. We could have a function like `logOnlyCommentInfo`. With our first approach of having the base trait, we'd have something like

```Scala
def logOnlyCommentInfo(info: BaseActivityInfo): Unit = info match {
  case 1 => ...
   ...
  case n =>
}
```
of course we could simplify this by using the wildcard operator `_` for pattern matching, but I think that could sometimes be error prone, with our approach of having a sealed trait up top, we could just have
```Scala
def logOnlyCommentInfo(info: CommentActivityInfo): Unit = info match {
  case 1 => ..
  case "last sub type of CommentActivityInfo" => ...
}
```
I would also argue that this is much safer. We get a compile-time error if we fail to exhaustively provide a case for all subtypes of `CommentActivityInfo`.

What are your thoughts?

takezoe replied:
> I think the benefit of this would be future ergonomics and extensibility.

If we have a plan to have such narrower logging methods, intermediate traits like `CommentActivityInfo` would be beneficial in the near future. I understand the technical benefit of it but I'm wondering whether there is such a possibility in our future.

Anyway, introducing intermediate traits doesn't have a downside while it may be beneficial in the future, so I'm not against introducing intermediate traits.

jyoo980 replied:
Decided to remove intermediate traits, see https://github.com/gitbucket/gitbucket/pull/2526#discussion_r478752552

takezoe commented on src/main/scala/gitbucket/core/model/activity/comment/CommentActivityInfo.scala:
```diff
@@ -0,0 +1,95 @@
+  def activityUserName: String
+  def comment: String
+  def maybeIssueId: Option[Int]
+  def maybeCommitId: Option[String]
```
Why do we need to have these `maybeXxxx` methods? It looks no one refer to them.

jyoo980 replied:
See https://github.com/gitbucket/gitbucket/pull/2526#discussion_r478525708

takezoe commented on src/main/scala/gitbucket/core/model/activity/tag/TagActivityInfo.scala:
```diff
@@ -0,0 +1,53 @@
+package gitbucket.core.model.activity.tag
```
I feel a little uncomfortable in the package structure of ActivityInfo classes. For example, `TagActivityInfo` and `BranchActivityInfo` should be categorized as `git` separated from `IssueActivityInfo`, etc. Also, `ForkActivityAction` should be categorized as `repository`.

I think simply putting all ActivityInfo classes under `activity` package would be fine because in other types of components like controllers or models also have a flat package structure.

jyoo980 replied:
That sounds good to me, I can put all the classes under the `activity` package, since this project appears to favour a flat package structure.

takezoe reviewed (commented):
@jyoo980 Overrall, this looks very good to me, yet I left some comments on the structure of ActivityInfo's type hierarchy and packages.

jyoo980 commented:
> @jyoo980 Overall, this looks very good to me, yet I left some comments about the structure of ActivityInfo's type hierarchy and packages.

Thanks so much for all your feedback, @takezoe, I've addressed your comments 🙂

jyoo980 commented on src/main/scala/gitbucket/core/model/activity/comment/CommentActivityInfo.scala:
```diff
@@ -0,0 +1,95 @@
+  issueId: Int
+) extends CommentActivityInfo {
+
+  override def maybeIssueId: Option[Int] = Some(issueId)
```
In reference to [your comment](https://github.com/gitbucket/gitbucket/pull/2526/files/18e3955e5b496d541220b72ae1b02a37d694688d#r478452455), a overrided value is provided here

takezoe replied:
My question was who uses this value? Is this also for future use?

jyoo980 replied:
Sorry for not addressing your question properly; this value is necessary for the `CommentActivityInfo` trait to be subtyped by both `IssueCommentInfo` and `CommitCommentInfo`. 

I see that you've also asked a question regarding why the intermediate trait is necessary, and I've provided an explanation [here](https://github.com/gitbucket/gitbucket/pull/2526#discussion_r478523839). In my opinion, this is a small sacrifice in order to have the benefits of exhaustive pattern-matching and compile-time checks

takezoe replied:
I understand your intention of introducing intermediate traits. Regarding these `maybeXxxx` methods, since no one calls these methods currently, so I thought these methods are defined for future use as same as intermediate traits. Is my understanding correct?

My personal preference is adding something when we need them. Therefore, these methods (and intermediate traits) looks redundant for me at the moment, to be honest.

jyoo980 replied:
Thanks for the feedback; I guess we have different ideas of how this should look. I'll defer to you since you have many more years of experience with this codebase than I have.

jyoo980 commented on src/main/scala/gitbucket/core/model/activity/comment/CommentActivityInfo.scala:
```diff
@@ -0,0 +1,95 @@
+
+  override def maybeIssueId: Option[Int] = None
+
+  override def maybeCommitId: Option[String] = Some(commitId)
```
In reference to [your comment](https://github.com/gitbucket/gitbucket/pull/2526/files/18e3955e5b496d541220b72ae1b02a37d694688d#r478452455), a overrided value is provided here

takezoe reviewed (approved):
Thanks for updating. Looks good to me. 👍
